### PR TITLE
Up: teardown when command context is cancelled

### PR DESCRIPTION
**Pull request by laurazard**
**What I did**

Previously, if a long-lived plugin process (such as an execution of `compose up`) was running and then detached from a terminal, signalling the parent CLI process to exit would leave the plugin process behind.

To address this, changes were introduced on the CLI side (see: https://github.com/docker/cli/pull/4599) to enable the CLI to notify a running plugin process that it should exit. This makes it so that, when the parent CLI process is going to exit, the command context of the plugin command being executed is cancelled.

This PR takes advantage of that change by using that channel on an `up` to know when to teardown.

Also fixed the "cancellable context" detection in `cmd/compose/compose.go`, as the string suffix matching was missing contexts such as:
```go
context.Background.WithCancel.WithValue(type trace.traceContextKeyType, val <not Stringer>).WithValue(type api.DryRunKey, val <not Stringer>)
```

*Before*:

https://github.com/docker/compose/assets/70572044/e1ffa1f5-469b-410a-a281-61aa970867bd

*After*:

https://github.com/docker/compose/assets/70572044/9e27f4da-4f02-4aef-addb-9c72e605a98a

**Related issue**
https://github.com/docker/cli/pull/4599

**(not mandatory) A picture of a cute animal, if possible in relation to what you did**

Screenshot 2023-12-21 at 12 15 46

### CODE REVIEW

**Comment by laurazard** [comment]
Looks like `TestComposeCancel/metrics_on_cancel_Compose_build` is failing, I'll take a look.

File: cmd/compose/compose.go [hunk 1]
```diff
@@ -277,7 +278,12 @@ const PluginName = "compose"
 
 // RunningAsStandalone detects when running as a standalone program
 func RunningAsStandalone() bool {
```
Comment (ndeloof) [inline cmd/compose/compose.go:280]: Should have been replaced by https://github.com/docker/cli/blob/ae41fd913eea86eb4be153782846c33935a1039e/cli-plugins/plugin/plugin.go#L223

Comment (laurazard) [reply cmd/compose/compose.go]: Agree! 😅

**Review by thaJeztah — APPROVED** [review]
LGTM

**Review by milas — APPROVED** [review]
LGTM, and you're my hero for fixing this!

I tested standalone, with Compose CLI trampoline, and with Docker CLI and all looks good. (Well, the standalone case does return exit code 130 but that's not a new regression; I'll open a quick PR for that later.)

**Pull request merged**